Fix entity offset double-conversion bug

j3h opened the pull request:
The entity offset conversion routines sometimes could double-convert
an entity, when its offset adjustment caused it to be examined twice.
This commit uses different routines for conversion in either
direction, that each take special care not to double-convert entities.

This bug was revealed when testing against real user data.

j3h committed:
Fix corner-case bugs in entity offset conversion

The entity offset conversion routines sometimes could double-convert
an entity, when its offset adjustment caused it to be examined twice.
This commit uses different routines for conversion in either
direction, that each take special care not to double-convert entities.

Commit:
Revert Extractor change.

Commit:
Upgraded to the latest conformance suite

Commit:
Refactor entity offset convertion.

keitaf commented:
Thank you for finding the bug and submitting a bug fix.

I tried to simplify your logic by removing nested while/for loops and having a common util function for both conversions.
https://github.com/twitter/twitter-text-java/compare/master...fix_utf16_index

Can you please check and see if it looks OK?

j3h committed:
UTF-16 conversion: Further simplify and document preconditions

j3h commented:
Looks good. As long as the list of entities is ordered by start index, this is much more efficient and easier to understand. I documented the precondition that the list of entities must be ordered, and switched back to iterating forward, since its easier to understand and using the iterator fixes the double-conversion bug.

keitaf commented:
Great! Yes actually there's no reason to iterate backward now.

Maybe we can sort entities by index in convertUnicodeIndices()?

```
Collections.<Entity>sort(entities, new Comparator<Entity>() {
  public int compare(Entity e1, Entity e2) {
    return e1.start - e2.start;
  }
});
```

keitaf commented:
Also I ported this onto twitter-text-js. Can you please also check this?
https://github.com/twitter/twitter-text-js/pull/45

j3h committed:
Another simplification of index conversion.

Remove the explicit iterator and lift the restriction that the
entities must be in a particular order. Also simplify logic.

keitaf commented:
LGTM! Love the Object-oriented style.
